### DISCUSSION TITLE
Implementing Social Login

### DISCUSSION THREAD

**Discussion by fugacityengr** [piccolo-orm/piccolo#440] [Q&A]
Hello, I very much like the batteries-included approach by this ORM compared to other ORMs that can be used with FastAPI.

However, I cannot seem to find discussions nor documentation for how to implement social logins (OAuth2) using Piccolo ORM.
Could anyone point me to an example or a discussion that I could follow to implement social login with Piccolo ORM?

**Comment by dantownsend** [comment]
@fugacityengr I've integrated with things like Google Sign-In with some of my own apps. There isn't really anything explicitly built into Piccolo to support it.

If you're trying to make the admin work with third party login, you could override the `login` method of `BaseUser`:

https://github.com/piccolo-orm/piccolo/blob/503bf245961f5d9e7b9612e2f5170246447cec21/piccolo/apps/user/tables.py#L153

For example:

```python
from piccolo.apps.user.tables import BaseUser

class CustomBaseUser(BaseUser):
    @classmethod
    async def login(cls, username: str, password: str) -> t.Optional[int]:
        # Call a third party endpoint, if valid return user ID
```

Then with Piccolo admin you can do:

```python
create_admin(tables=[Table1, Table2], auth_table=CustomBaseUser)
```

Is that the kind of thing you're looking for?

**Reply by fugacityengr** [reply]
@dantownsend Thank you for answering.
Yes something like that.
I am trying to implement social login alongside a normal login (the usual username + password combo).

**Reply by sinisaos** [reply]
@fugacityengr It's more in the scope of FastAPI than Piccolo, but you can try the [Authlib](https://docs.authlib.org/en/latest/) library. [Here](https://github.com/authlib/demo-oauth-client/tree/master/fastapi-google-login) is an example of Google login (**important!!!** you must `pip install Authlib==1.0.0rc1` because PyPI version (`Authlib 0.15.5`) does not work for me) . You can create a new Piccolo asgi app and then in `app.py` you can add code from the Authlib example. Once you get a user from Google modify `homepage` method to this
```python
# from piccolo.apps.user.tables import BaseUser 
# and all other imports and code

@app.get("/")
async def homepage(request: Request):
    user = request.session.get("user")
    if user:
        # if user is not in db save it
        if not (
            await BaseUser.exists()
            .where(BaseUser.email == user["email"])
            .run()
            or await BaseUser.exists()
            .where(BaseUser.username == user["given_name"])
            .run()
        ):
            await BaseUser(
                username=user["given_name"], email=user["email"]
            ).save().run()
        # if user is in db just show google user
        html = f"<pre>{user}</pre> <a href='/logout'>Logout {user['given_name']}</a>"
        return HTMLResponse(html)
    return HTMLResponse('<a href="/login">Login</a>')
```
With this method you save Google user to Piccolo BaseUser and later use it as a normal BaseUser. I don’t know it’s a good approach, but it works. I will try to implement a social login in this [project template](https://github.com/sinisaos/piccolo-vue-starter), but it uses Vue as a frontend and I have to think about how to do it.
Hope this help you.

**Reply by dantownsend** [reply]
Thanks for this - Authlib looks like a great library.
